js/web/lib/wasm/jsep/webgpu/ops/instance-norm.ts

```diff
@@ -26,22 +26,25 @@ const createInstanceNormProgramInfo =
       const axis = 2;
       const normCount = ShapeUtil.sizeToDimension(xShape, axis);
       const normSize = ShapeUtil.sizeFromDimension(xShape, axis);
+      const components = getMaxComponents(normSize);
+      const normPackedSize = normSize / components;
       const C = xShape[1];
-      const x = inputVariable('x', inputs[0].dataType, [xShape[0], xShape[1], normSize]);
+      const x = inputVariable('x', inputs[0].dataType, [xShape[0], xShape[1], normPackedSize], components);
       const scale = inputVariable('scale', inputs[1].dataType, inputs[1].dims);
       const bias = inputVariable('bias', inputs[2].dataType, inputs[2].dims);
-      const output = outputVariable('output', inputs[0].dataType, [xShape[0], xShape[1], normSize]);
+      const output = outputVariable('output', inputs[0].dataType, [xShape[0], xShape[1], normPackedSize], components);
       const variables = [x, scale, bias, output];
       const dataType = x.type.value;
+      const f32Type = components === 1 ? 'f32' : `vec${components}<f32>`;
       const workgroupSize = 64;
       const getShaderSource = (shaderHelper: ShaderHelper) => `
 
   const C: u32 = ${C};
   const normSize: u32 = ${normSize};
   const epsilon: f32 = ${attributes.epsilon};
-  var<workgroup> meanShared : ${dataType};
-  var<workgroup> squaredNormShared : ${dataType};
-  var<workgroup> workgroupShared : array<${dataType}, ${workgroupSize}>;
+  var<workgroup> meanShared : f32;
```

Comment:
would we not want to use ${dataType} ?

Reply:
No, here we must use f32 since `dataType` may be f16 which may results the result overflow. See the definition of `SkipLayerNormalization` in https://github.com/microsoft/onnxruntime/blob/main/docs/ContribOperators.md#com.microsoft.SkipLayerNormalization, the `mean` and `inv_std_var` 's type is `U : tensor(float)` , which constrain mean and inv_std_var to float tensors. So you can see that `NHWC` InstanceNorm, `LayerNorm` and `LayerNorm` all use f32 to store `mean` and `inv_std_var`.

Reply:
got it